{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "420d448b-a1e7-4a7d-aa3c-5ee22f2a7232",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Goal:Check acz vessel_size_gradient\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "587dcd50-1490-4d93-8f1d-b23d0c6c1b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import os \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from functions import compare_obs_values_within_groups_to_excel\n",
    "import matplotlib.pyplot as plt\n",
    "import palantir\n",
    "# from statannotations.Annotator import Annotator\n",
    "\n",
    "adata_name='venous_ec'\n",
    "figures = \"data/figures/figures/acz_iscience_vessel_size\"\n",
    "data = \"data/single_cell_files/scanpy_files\"\n",
    "\n",
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cabe2535-cfde-4ba2-9fdf-c41aa6593ba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('/home/carsten/alvira_bioinformatics/data/alvira_sc_data/geo_versions/perinatal_lung.h5ad')\n",
    "sc.pp.normalize_total(adata,target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "adata.layers[\"log1p\"] = adata.X.copy()\n",
    "sorted(adata.obs['Cell Subtype'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30fc661d-9bbd-4e6b-85e7-9b572cb72374",
   "metadata": {},
   "outputs": [],
   "source": [
    "## read in cell cycle genes and score\n",
    "cell_cycle_genes = [x.strip() for x in open('/home/carsten/alvira_bioinformatics/venous_ec_scrnaseq/data/outside_data/regev_lab_cell_cycle_genes.txt')]\n",
    "cell_cycle_genes = [x.lower().capitalize() for x in cell_cycle_genes]\n",
    "s_genes = cell_cycle_genes[:43]\n",
    "g2m_genes = cell_cycle_genes[43:]\n",
    "cell_cycle_genes = [x for x in cell_cycle_genes if x in adata.var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66f51d19-e36e-4d16-b051-7cb287570627",
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_cts = ['Arterial EC I',\n",
    " 'Arterial EC II', 'Early Car4- capillaries','Late Car4- capillaries',\n",
    "'Venous EC','Proliferative EC', 'Proliferative venous EC',\n",
    "]\n",
    "endo_adata = adata[adata.obs['Cell Subtype'].isin(endo_cts)].copy()\n",
    "sc.tl.score_genes_cell_cycle(endo_adata, s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "sc.pp.regress_out(endo_adata, ['S_score', 'G2M_score'])\n",
    "endo_adata.layers['cc_regress'] = endo_adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14df590a-614d-4e7f-97c5-8983d30c0712",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pp.highly_variable_genes(endo_adata,batch_key='Mousename')\n",
    "sc.pp.pca(endo_adata, mask_var=\"highly_variable\")\n",
    "sce.pp.harmony_integrate(endo_adata,'Mousename',max_iter_harmony = 20)\n",
    "sc.pp.neighbors(endo_adata, use_rep=\"X_pca_harmony\")\n",
    "sc.tl.leiden(endo_adata, key_added=\"leiden\")\n",
    "endo_adata.obs['Cell Subtype 2'] = endo_adata.obs['leiden'].map({'0':'Cap1','1':'Cap1','2':'Cap1','3':'Venous EC','4':'Cap1','5':'Arterial EC','6':'Cap1','7':'Cap1','8':'Cap1','9':'Cap1','10':'Arterial EC',})\n",
    "endo_adata.uns['Cell Subtype 2_colors']= ['#4A90E2','#9B59B6','#E35D6A']\n",
    "sc.tl.umap(endo_adata,min_dist=0.5)\n",
    "endo_adata.X = endo_adata.layers['log1p'].copy()\n",
    "sc.tl.rank_genes_groups(endo_adata,'leiden',method='wilcoxon')\n",
    "sc.pl.rank_genes_groups_dotplot(endo_adata,dendrogram=False)\n",
    "\n",
    "for color in ['Cell Subtype','Cell Subtype 2','Timepoint','Treatment','leiden','Gja5','Slc6a2','Kit','Hey1','Nr2f2','Mecom','Eln','Mgp','Col4a1','Col4a2']:\n",
    "    sc.pl.umap(endo_adata,color=color,use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082ff7f-bcd3-43f4-af14-e9f0370bf976",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(endo_adata,color='Cell Subtype 2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "398dbb71-77e3-4616-bec0-126afc766c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "import palantir\n",
    "import cellrank as cr\n",
    "import scvelo as scv\n",
    "\n",
    "root_ct = 'Cap1'\n",
    "terminal_cts = ['Arterial EC','Venous EC']\n",
    "celltype='Cell Subtype 2'\n",
    "\n",
    "palantir.utils.run_diffusion_maps(endo_adata,\n",
    "                                           n_components=5)\n",
    "fig = palantir.plot.plot_diffusion_components(endo_adata)[0]\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/aczpalantir_diffusion_components.png')\n",
    "plt.close()\n",
    "palantir.utils.determine_multiscale_space(endo_adata)\n",
    "\n",
    "palantir.utils.run_magic_imputation(endo_adata)\n",
    "subset = endo_adata[endo_adata.obs[celltype] == root_ct]\n",
    "umap1_values = subset.obsm['X_umap'][:, 1]\n",
    "min_idx = np.argmin(umap1_values)\n",
    "root_cell = subset.obs_names[min_idx]\n",
    "terminal_states = []\n",
    "for ct in terminal_cts:\n",
    "    subset = endo_adata[endo_adata.obs[celltype] == ct]\n",
    "    if ct =='Arterial EC':\n",
    "        # Get the index (obs_names) of the cell with the min UMAP1 (usually component 0)\n",
    "        umap1_values = subset.obsm['X_umap'][:, 1]\n",
    "        max_idx = np.argmin(umap1_values)\n",
    "        # Return the cell name\n",
    "        terminal_states.append(subset.obs_names[max_idx])\n",
    "    else:\n",
    "        umap1_values = subset.obsm['X_umap'][:, 0]\n",
    "        max_idx = np.argmin(umap1_values)\n",
    "        # Return the cell name\n",
    "        terminal_states.append(subset.obs_names[max_idx])\n",
    "        \n",
    "terminal_states = pd.Series(index=terminal_states, data=terminal_cts, dtype='object')\n",
    "\n",
    "fig = palantir.plot.highlight_cells_on_umap(endo_adata, [root_cell]+terminal_states)[0]\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/aczpalantir_terminal_cells.png')\n",
    "plt.close()\n",
    "\n",
    "palantir.core.run_palantir(\n",
    "    endo_adata, root_cell, num_waypoints=500, terminal_states=terminal_states\n",
    ")\n",
    "\n",
    "fig = palantir.plot.plot_palantir_results(endo_adata, s=3)\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/aczpalantir_results.png')\n",
    "plt.close()\n",
    "iroot = endo_adata.obs.index.get_loc(root_cell)\n",
    "endo_adata.uns[\"iroot\"] = iroot\n",
    "sc.tl.dpt(endo_adata)\n",
    "\n",
    "try:\n",
    "    palantir.presults.select_branch_cells(endo_adata, q=.01, eps=.01,pseudo_time_key='dpt_pseudotime')\n",
    "\n",
    "    fig = palantir.plot.plot_branch_selection(endo_adata)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(f'{figures}/aczpalantir_branch_selection.png')\n",
    "    plt.close()\n",
    "\n",
    "except:\n",
    "    pass\n",
    "\n",
    "sc.tl.diffmap(endo_adata)\n",
    "scv.pl.scatter(\n",
    "    endo_adata,\n",
    "    basis=\"diffmap\",\n",
    "    c=[celltype, iroot],\n",
    "    legend_loc=\"right\",\n",
    "    components=[\"2, 3\"],\n",
    "    show=False,\n",
    "    save=f'aczdiffmap_{celltype}_root_cell.png'\n",
    ")\n",
    "\n",
    "\n",
    "sc.pl.embedding(\n",
    "    endo_adata,\n",
    "    basis=\"umap\",\n",
    "    color=[\"dpt_pseudotime\", \"palantir_pseudotime\"],\n",
    "    color_map=\"viridis\",\n",
    "    show=False,\n",
    "    save='_acz_pseudotimes.png'\n",
    ")\n",
    "\n",
    "palantir.presults.compute_gene_trends(\n",
    "    endo_adata,\n",
    "    expression_key=\"MAGIC_imputed_data\",\n",
    "    pseudo_time_key='dpt_pseudotime'\n",
    ")\n",
    "\n",
    "pk = cr.kernels.PseudotimeKernel(endo_adata, time_key=\"palantir_pseudotime\")\n",
    "pk.compute_transition_matrix()\n",
    "pk.plot_projection(basis=\"umap\", color=celltype, recompute=True,legend_loc='right margin',\n",
    "                         save=f'{figures}/aczpalantir_pseudotime_stream.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bed69eb-2eea-471d-a12f-c49009303722",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "def correlate_genes_with_pseudotime(adata, layer=None, method='spearman',pseudotime='dpt_pseudotime'):\n",
    "    \"\"\"\n",
    "    Correlates all genes with pseudotime in an AnnData object.\n",
    "\n",
    "    Parameters:\n",
    "    - adata: AnnData object with pseudotime in `adata.obs['pseudotime']`\n",
    "    - layer: (Optional) Layer to use instead of adata.X (e.g., 'log1p', 'counts')\n",
    "    - method: Correlation method, either 'spearman' (default) or 'pearson'\n",
    "\n",
    "    Returns:\n",
    "    - pandas DataFrame with genes as index and columns: ['correlation', 'pval']\n",
    "    \"\"\"\n",
    "    if pseudotime not in adata.obs:\n",
    "        raise ValueError(\"Pseudotime must be stored in adata.obs['pseudotime'].\")\n",
    "\n",
    "    # Get expression matrix\n",
    "    X = adata.X if layer is None else adata.layers[layer]\n",
    "    if not isinstance(X, pd.DataFrame):\n",
    "        X = pd.DataFrame(X.toarray() if hasattr(X, \"toarray\") else X,\n",
    "                         index=adata.obs_names, columns=adata.var_names)\n",
    "\n",
    "    # Extract pseudotime\n",
    "    pseudotime = adata.obs[pseudotime]\n",
    "\n",
    "    # Run correlation\n",
    "    results = []\n",
    "    for gene in X.columns:\n",
    "        if method == 'spearman':\n",
    "            corr, pval = spearmanr(X[gene], pseudotime)\n",
    "        elif method == 'pearson':\n",
    "            corr, pval = X[gene].corr(pseudotime), None  # Pearson p-value not computed here\n",
    "        else:\n",
    "            raise ValueError(\"Method must be 'spearman' or 'pearson'.\")\n",
    "        results.append((gene, corr, pval))\n",
    "\n",
    "    result_df = pd.DataFrame(results, columns=['gene', 'correlation', 'pval']).set_index('gene')\n",
    "    return result_df.sort_values('correlation', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83a31b9-6dee-4e76-bf7f-0bb11cd42b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_dfs = {}\n",
    "for ct in ['Arterial EC','Venous EC']:\n",
    "    # ct_adata = endo_adata[endo_adata.obsm['branch_masks'][ct]]    \n",
    "    ct_adata = endo_adata[endo_adata.obs['Cell Subtype 2']==ct]\n",
    "\n",
    "    df = correlate_genes_with_pseudotime(ct_adata,method='pearson',pseudotime='palantir_pseudotime')\n",
    "    corr_dfs[ct]=df.dropna(how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47d9a07-3669-4a2f-a92b-c480e832d6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_n_genes=50\n",
    "arterial_large_genes = corr_dfs['Arterial EC'].head(top_n_genes).index.tolist()\n",
    "venous_large_genes = corr_dfs['Venous EC'].head(top_n_genes).index.tolist()\n",
    "arterial_small_genes = corr_dfs['Arterial EC'].tail(top_n_genes).index.tolist()[::-1]\n",
    "venous_small_genes = corr_dfs['Venous EC'].tail(top_n_genes).index.tolist()[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d093c1f3-3e97-43e5-8413-93db1c27da6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2,venn3\n",
    "\n",
    "# Define your lists\n",
    "\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn = venn2([set(arterial_large_genes), set(venous_large_genes)], \n",
    "             set_labels=('Arterial', 'Venous'), \n",
    "             set_colors=('#4A90E2', '#E35D6A'), \n",
    "             alpha=0.7)\n",
    "\n",
    "# Optional: Customize font size\n",
    "for text in venn.set_labels:\n",
    "    text.set_fontsize(12)\n",
    "for text in venn.subset_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(12)\n",
    "\n",
    "# Show the plot\n",
    "plt.title(\"Top 50 genes positively correlated with pseudotime\")\n",
    "plt.savefig(f'{figures}/aczvenn_diagram_large.png',dpi=300,bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b9f29fd-f468-4f62-8595-ab6bb5430268",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2\n",
    "\n",
    "# Define your lists\n",
    "\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn = venn2([set(arterial_small_genes), set(venous_small_genes)], \n",
    "             set_labels=('Arterial', 'Venous'), \n",
    "             set_colors=('#4A90E2', '#E35D6A'), \n",
    "             alpha=0.7)\n",
    "\n",
    "# Optional: Customize font size\n",
    "for text in venn.set_labels:\n",
    "    text.set_fontsize(12)\n",
    "for text in venn.subset_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(12)\n",
    "\n",
    "# Show the plot\n",
    "plt.title(\"Top 50 genes negatively correlated with pseudotime\")\n",
    "plt.savefig(f'{figures}/aczvenn_diagram_small.png',dpi=300,bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3478a58-c6ec-48f7-b126-052174fa4f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_genes = [x for x in arterial_large_genes if x in venous_large_genes]\n",
    "small_genes = [x for x in arterial_small_genes if x in venous_small_genes]\n",
    "sc.tl.score_genes(endo_adata,large_genes,score_name='large_score')\n",
    "sc.tl.score_genes(endo_adata,small_genes,score_name='small_score')\n",
    "endo_adata.obs['Vessel size score'] = endo_adata.obs['large_score'] - endo_adata.obs['small_score']\n",
    "\n",
    "\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler()\n",
    "def normalize_dataframe(df):\n",
    "    # Initialize the MinMaxScaler\n",
    "    scaler = MinMaxScaler(feature_range=(0, 1))\n",
    "    # Fit the scaler on the data and transform each column\n",
    "    df_normalized = pd.DataFrame(scaler.fit_transform(df), index=df.index, columns=df.columns)\n",
    "    return df_normalized\n",
    "endo_adata.obs['Vessel size score'] = scaler.fit_transform(endo_adata.obs[['Vessel size score']])\n",
    "endo_adata.obs['Vessel size category'] = pd.cut(endo_adata.obs['Vessel size score'], bins=4,labels=['capillary','small','medium','large'])\n",
    "sc.pl.umap(endo_adata,color=['Vessel size score'],cmap='Oranges',size=size,frameon=False,save='_aczvessel_size_score.png')\n",
    "sc.pl.umap(endo_adata,color=['Vessel size category'],cmap='viridis',size=size,frameon=False,save='_aczvessel_size_category.png')\n",
    "sc.pl.umap(endo_adata,color=['Cell Subtype 2'],cmap='viridis',size=size,legend_loc='on data',legend_fontsize=10, legend_fontoutline=1,frameon=False,save='aczcellsubtype.png')\n",
    "sc.pl.umap(endo_adata,color=['Mgp'],cmap='viridis',size=size,frameon=False,save='aczmgp.png')\n",
    "sc.pl.umap(endo_adata,color=['Col4a1'],cmap='viridis',size=size,frameon=False,save='aczcol4a1.png')\n",
    "sc.pl.umap(endo_adata,color=['Col4a2'],cmap='viridis',size=size,frameon=False,save='aczcol4a2.png')\n",
    "sc.pl.umap(endo_adata,color=['Eln'],cmap='viridis',size=size,frameon=False,save='aczeln.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb3864e-37ed-4ecf-850f-faafd1c30cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(endo_adata,color = large_genes + small_genes,cmap='viridis',hspace=0.5,save='acz_allsize.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f72771-68fe-4422-a94c-46301f4e33f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_adata.write(f'{figures}/vessel_size.gz.h5ad',compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d4cffc-b365-4579-9d64-d5a87013b3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1,1,figsize=(2,2))\n",
    "palantir.plot.plot_trajectory(\n",
    "    endo_adata, # your anndata\n",
    "    \"Arterial EC\", # the branch to plot\n",
    "    cell_color=\"dpt_pseudotime\", # the ad.obs colum to color the cells by\n",
    "    n_arrows=5, # the number of arrow heads along the path\n",
    "    color='#4A90E2', # the color of the path and arrow heads\n",
    "    scanpy_kwargs=dict(cmap=\"viridis\",size=size), # arguments passed to scanpy.pl.embedding\n",
    "    arrowprops=dict(arrowstyle=\"->,head_length=.25,head_width=.25\", lw=2), # appearance of the arrow heads\n",
    "    lw=2, # thickness of the path\n",
    "ax=ax\n",
    "    # pseudotime_interval=(0, .9), # interval of the pseudotime to cover with the path\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(f'{figures}/aczpalantir_art_trajectory.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a8b44ed-91b8-4545-954d-daa9e41548d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1,1,figsize=(2,2))\n",
    "palantir.plot.plot_trajectory(\n",
    "    endo_adata, # your anndata\n",
    "    \"Venous EC\", # the branch to plot\n",
    "    cell_color=\"dpt_pseudotime\", # the ad.obs colum to color the cells by\n",
    "    n_arrows=5, # the number of arrow heads along the path\n",
    "    color='#E35D6A', # the color of the path and arrow heads\n",
    "    scanpy_kwargs=dict(cmap=\"viridis\",size=size), # arguments passed to scanpy.pl.embedding\n",
    "    arrowprops=dict(arrowstyle=\"->,head_length=.25,head_width=.25\", lw=2), # appearance of the arrow heads\n",
    "    lw=2, # thickness of the path\n",
    "ax=ax\n",
    "    # pseudotime_interval=(0, .9), # interval of the pseudotime to cover with the path\n",
    ")\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/aczpalantir_ven_trajectory.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef11e561-8348-4749-81ab-7cda2ab14271",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(endo_adata,['Cxcl12','Cxcr4','Ackr3'],groupby=['Cell Subtype 2','Vessel size category','Timepoint','Treatment']).add_totals().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b52cdc-0f18-449d-9249-d04bf30c5cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1,1))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15\n",
    "for gene in ['Cxcl12','Cxcr4','Ackr3','Esr2','Dkk2','Moxd1']:\n",
    "    sc.pl.umap(endo_adata,color=[gene],cmap='viridis',size=size,frameon=False,save=f'acz_{gene}.png')\n",
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3c93955-724e-4bc9-81ad-b9fc4bd4a224",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls = []\n",
    "for x,y in zip(endo_adata.obs['Cell Subtype 2'],endo_adata.obs['Vessel size category']):\n",
    "    if x =='Cap1':\n",
    "        ls.append('Cap1')\n",
    "    else:\n",
    "        if y =='capillary':\n",
    "            ls.append(f'Cap1')\n",
    "            continue\n",
    "        if x =='Arterial EC':\n",
    "            x = 'PAEC'\n",
    "        else:\n",
    "            x ='PVEC'\n",
    "        ls.append(f'{x} {y[0].upper()}')\n",
    "endo_adata.obs['ct_s'] = ls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48e54e20-b630-488c-9857-3e4aa6b9c4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "\n",
    "gene_ls = ['Dkk2','Sox6',\n",
    "           'Lama3','Esr2',\n",
    "           'Stc1',\n",
    "             'Glp1r','Kit','Wif1','Car2',\n",
    "           'Rarb','Chrm3',\n",
    "           'Gria3',\n",
    "           'Ptger3','Moxd1',\n",
    "             ]\n",
    "sc.pl.dotplot(endo_adata,gene_ls,groupby='ct_s',cmap='viridis',categories_order=['PAEC L','PAEC M','PAEC S','Cap1','PVEC S','PVEC M', 'PVEC L'],\n",
    "              standard_scale='var',save='size_axis.png')\n",
    "sc.pl.umap(endo_adata,color=gene_ls,hspace=0.5,wspace=0.3,cmap='viridis',save='size_axis.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fabd8a4-77eb-47c9-b8ad-fceee5d737cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ls = ['Ccdc85a','Glp1r','Sema3c',\n",
    "         'Sox4','Nrp1','Ifitm3',\n",
    "         'Ptprr','Adgrg6','Foxo1',\n",
    "         'Mgp','Eln','Nr4a2',\n",
    "]\n",
    "\n",
    "sc.pl.MatrixPlot(endo_adata,gene_ls[::-1],groupby='ct_s',categories_order=['PAEC L','PAEC M','PAEC S','Cap1','PVEC S','PVEC M', 'PVEC L'],\n",
    "              standard_scale='var').style(cmap='viridis').swap_axes().savefig(f'{figures}/matrixplot_size_markers.png',dpi=300,bbox_inches='tight')\n",
    "sc.pl.DotPlot(endo_adata,gene_ls[::-1],groupby='ct_s',categories_order=['PAEC L','PAEC M','PAEC S','Cap1','PVEC S','PVEC M', 'PVEC L'],\n",
    "              standard_scale='var').style(cmap='viridis').savefig(f'{figures}/matrixplot_size_markers.png',dpi=300,bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed59e169-b667-4bc7-a3e7-e552f11ab158",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
