{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "420d448b-a1e7-4a7d-aa3c-5ee22f2a7232",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Goal:Check hurskainen vessel_size_gradient\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "587dcd50-1490-4d93-8f1d-b23d0c6c1b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import os \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from functions import compare_obs_values_within_groups_to_excel\n",
    "import matplotlib.pyplot as plt\n",
    "import palantir\n",
    "# from statannotations.Annotator import Annotator\n",
    "\n",
    "adata_name='venous_ec'\n",
    "figures = \"data/figures/figures/hurskainen\"\n",
    "data = \"data/single_cell_files/scanpy_files\"\n",
    "\n",
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20eba6ba-524c-40df-a9b7-6e9d1ed4f4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_adata= sc.read(f'{figures}/vessel_size.gz.h5ad')\n",
    "sc.pl.umap(endo_adata,color=['Cell Subtype'],cmap='viridis',size=size,legend_fontsize=10, legend_fontoutline=1,frameon=False,save='hurskainencellsubtype.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cabe2535-cfde-4ba2-9fdf-c41aa6593ba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = '/home/carsten/alvira_bioinformatics/data/external_datasets/hurskainen2021_hyperoxia_lung/raw'\n",
    "metadata = pd.read_csv(os.path.join(source, \"GSE151974_cell_metadata_postfilter.csv.gz\"),\n",
    "                       compression='gzip', \n",
    "                       header=0,\n",
    "                       index_col= 0,\n",
    "                       sep=',', \n",
    "                      )\n",
    "\n",
    "adata = sc.read_csv(os.path.join(source,\"GSE151974_raw_umi_matrix_postfilter.csv.gz\"),\n",
    "                       first_column_names=True, \n",
    "                       delimiter=',', \n",
    "                    ).T\n",
    "adata.obs = metadata\n",
    "adata.obs.rename(columns = {'Age':'Timepoint',\n",
    "                  'Oxygen': 'Treatment',\n",
    "                  'CellType': 'celltype'},\n",
    "                 inplace = True)\n",
    "\n",
    "adata.obs['celltype'].replace({'Cap':'Cap1','Cap-a':'Cap2'},inplace=True)\n",
    "adata.obs['timepoint_treatment'] =adata.obs['Timepoint'].astype(str) + '_' + adata.obs['Treatment'].astype(str)\n",
    "sc.pp.calculate_qc_metrics(\n",
    "        adata,\n",
    "        expr_type=\"counts\",\n",
    "        log1p=True,\n",
    "        inplace=True,\n",
    "    )\n",
    "## read in cell cycle genes and score\n",
    "cell_cycle_genes = [x.strip() for x in open('/home/carsten/alvira_bioinformatics/venous_ec_scrnaseq/data/outside_data/regev_lab_cell_cycle_genes.txt')]\n",
    "cell_cycle_genes = [x.lower().capitalize() for x in cell_cycle_genes]\n",
    "s_genes = cell_cycle_genes[:43]\n",
    "g2m_genes = cell_cycle_genes[43:]\n",
    "cell_cycle_genes = [x for x in cell_cycle_genes if x in adata.var_names]\n",
    "sc.pp.normalize_total(adata, key_added=None, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.tl.score_genes(adata, ['Mki67', 'Top2a', 'Birc5', 'Hmgb2', 'Cenpf'], score_name='proliferation_score')\n",
    "adata.layers[\"log1p\"] = adata.X.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66f51d19-e36e-4d16-b051-7cb287570627",
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_cts = ['Art','Cap1',\n",
    "'Vein',\n",
    "]\n",
    "endo_adata = adata[adata.obs['celltype'].isin(endo_cts)].copy()\n",
    "sc.tl.score_genes_cell_cycle(endo_adata, s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "sc.pp.regress_out(endo_adata, ['S_score', 'G2M_score'])\n",
    "endo_adata.layers['cc_regress'] = endo_adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14df590a-614d-4e7f-97c5-8983d30c0712",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pp.highly_variable_genes(endo_adata,batch_key='orig.ident')\n",
    "sc.pp.pca(endo_adata, mask_var=\"highly_variable\")\n",
    "sce.pp.harmony_integrate(endo_adata,'orig.ident',max_iter_harmony = 20)\n",
    "sc.pp.neighbors(endo_adata, use_rep=\"X_pca_harmony\")\n",
    "sc.tl.leiden(endo_adata, key_added=\"leiden\")\n",
    "endo_adata.obs['Cell Subtype'] = endo_adata.obs['leiden'].map({'0':'Cap1','1':'Cap1','2':'Cap1','3':'Cap1','4':'Cap1','5':'Arterial EC','6':'Venous EC','7':'Cap1',})\n",
    "endo_adata.uns['Cell Subtype_colors']= ['#4A90E2','#9B59B6','#E35D6A']\n",
    "\n",
    "sc.tl.umap(endo_adata,min_dist=0.5)\n",
    "sc.tl.rank_genes_groups(endo_adata,'leiden',method='wilcoxon')\n",
    "sc.pl.rank_genes_groups_dotplot(endo_adata,dendrogram=False)\n",
    "\n",
    "for color in ['celltype','Timepoint','Treatment','leiden','Gja5','Slc6a2','Kit','Hey1','Nr2f2','Mecom','Eln','Mgp','Col4a1','Col4a2']:\n",
    "    sc.pl.umap(endo_adata,color=color,use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5eba609-d551-4181-8a50-81ba9dbec320",
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_adata.X = endo_adata.layers['log1p'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082ff7f-bcd3-43f4-af14-e9f0370bf976",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(endo_adata,color='Cell Subtype')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "398dbb71-77e3-4616-bec0-126afc766c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "import palantir\n",
    "import cellrank as cr\n",
    "import scvelo as scv\n",
    "\n",
    "root_ct = 'Cap1'\n",
    "terminal_cts = ['Arterial EC','Venous EC']\n",
    "celltype='Cell Subtype'\n",
    "\n",
    "palantir.utils.run_diffusion_maps(endo_adata,\n",
    "                                           n_components=5)\n",
    "fig = palantir.plot.plot_diffusion_components(endo_adata)[0]\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/hurskainenpalantir_diffusion_components.png')\n",
    "plt.close()\n",
    "palantir.utils.determine_multiscale_space(endo_adata)\n",
    "\n",
    "palantir.utils.run_magic_imputation(endo_adata)\n",
    "subset = endo_adata[endo_adata.obs[celltype] == root_ct]\n",
    "umap1_values = subset.obsm['X_umap'][:, 1]\n",
    "min_idx = np.argmin(umap1_values)\n",
    "root_cell = subset.obs_names[min_idx]\n",
    "terminal_states = []\n",
    "for ct in terminal_cts:\n",
    "    subset = endo_adata[endo_adata.obs[celltype] == ct]\n",
    "    if ct =='Arterial EC':\n",
    "        # Get the index (obs_names) of the cell with the min UMAP1 (usually component 0)\n",
    "        umap1_values = subset.obsm['X_umap'][:, 1]\n",
    "        max_idx = np.argmax(umap1_values)\n",
    "        # Return the cell name\n",
    "        terminal_states.append(subset.obs_names[max_idx])\n",
    "    else:\n",
    "        umap1_values = subset.obsm['X_umap'][:, 0]\n",
    "        max_idx = np.argmax(umap1_values)\n",
    "        # Return the cell name\n",
    "        terminal_states.append(subset.obs_names[max_idx])\n",
    "        \n",
    "terminal_states = pd.Series(index=terminal_states, data=terminal_cts, dtype='object')\n",
    "\n",
    "fig = palantir.plot.highlight_cells_on_umap(endo_adata, [root_cell]+terminal_states)[0]\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/hurskainenpalantir_terminal_cells.png')\n",
    "plt.close()\n",
    "\n",
    "palantir.core.run_palantir(\n",
    "    endo_adata, root_cell, num_waypoints=500, terminal_states=terminal_states\n",
    ")\n",
    "\n",
    "fig = palantir.plot.plot_palantir_results(endo_adata, s=3)\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/hurskainenpalantir_results.png')\n",
    "plt.close()\n",
    "iroot = endo_adata.obs.index.get_loc(root_cell)\n",
    "endo_adata.uns[\"iroot\"] = iroot\n",
    "sc.tl.dpt(endo_adata)\n",
    "\n",
    "try:\n",
    "    palantir.presults.select_branch_cells(endo_adata, q=.01, eps=.01,pseudo_time_key='dpt_pseudotime')\n",
    "\n",
    "    fig = palantir.plot.plot_branch_selection(endo_adata)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(f'{figures}/hurskainenpalantir_branch_selection.png')\n",
    "    plt.close()\n",
    "\n",
    "except:\n",
    "    pass\n",
    "\n",
    "sc.tl.diffmap(endo_adata)\n",
    "scv.pl.scatter(\n",
    "    endo_adata,\n",
    "    basis=\"diffmap\",\n",
    "    c=[celltype, iroot],\n",
    "    legend_loc=\"right\",\n",
    "    components=[\"2, 3\"],\n",
    "    show=False,\n",
    "    save=f'hurskainendiffmap_{celltype}_root_cell.png'\n",
    ")\n",
    "\n",
    "\n",
    "sc.pl.embedding(\n",
    "    endo_adata,\n",
    "    basis=\"umap\",\n",
    "    color=[\"dpt_pseudotime\", \"palantir_pseudotime\"],\n",
    "    color_map=\"viridis\",\n",
    "    show=False,\n",
    "    save='_hurskainen_pseudotimes.png'\n",
    ")\n",
    "\n",
    "palantir.presults.compute_gene_trends(\n",
    "    endo_adata,\n",
    "    expression_key=\"MAGIC_imputed_data\",\n",
    "    pseudo_time_key='dpt_pseudotime'\n",
    ")\n",
    "\n",
    "pk = cr.kernels.PseudotimeKernel(endo_adata, time_key=\"palantir_pseudotime\")\n",
    "pk.compute_transition_matrix()\n",
    "pk.plot_projection(basis=\"umap\", color=celltype, recompute=True,legend_loc='right margin',\n",
    "                         save=f'{figures}/hurskainenpalantir_pseudotime_stream.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bed69eb-2eea-471d-a12f-c49009303722",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "def correlate_genes_with_pseudotime(adata, layer=None, method='spearman',pseudotime='dpt_pseudotime'):\n",
    "    \"\"\"\n",
    "    Correlates all genes with pseudotime in an AnnData object.\n",
    "\n",
    "    Parameters:\n",
    "    - adata: AnnData object with pseudotime in `adata.obs['pseudotime']`\n",
    "    - layer: (Optional) Layer to use instead of adata.X (e.g., 'log1p', 'counts')\n",
    "    - method: Correlation method, either 'spearman' (default) or 'pearson'\n",
    "\n",
    "    Returns:\n",
    "    - pandas DataFrame with genes as index and columns: ['correlation', 'pval']\n",
    "    \"\"\"\n",
    "    if pseudotime not in adata.obs:\n",
    "        raise ValueError(\"Pseudotime must be stored in adata.obs['pseudotime'].\")\n",
    "\n",
    "    # Get expression matrix\n",
    "    X = adata.X if layer is None else adata.layers[layer]\n",
    "    if not isinstance(X, pd.DataFrame):\n",
    "        X = pd.DataFrame(X.toarray() if hasattr(X, \"toarray\") else X,\n",
    "                         index=adata.obs_names, columns=adata.var_names)\n",
    "\n",
    "    # Extract pseudotime\n",
    "    pseudotime = adata.obs[pseudotime]\n",
    "\n",
    "    # Run correlation\n",
    "    results = []\n",
    "    for gene in X.columns:\n",
    "        if method == 'spearman':\n",
    "            corr, pval = spearmanr(X[gene], pseudotime)\n",
    "        elif method == 'pearson':\n",
    "            corr, pval = X[gene].corr(pseudotime), None  # Pearson p-value not computed here\n",
    "        else:\n",
    "            raise ValueError(\"Method must be 'spearman' or 'pearson'.\")\n",
    "        results.append((gene, corr, pval))\n",
    "\n",
    "    result_df = pd.DataFrame(results, columns=['gene', 'correlation', 'pval']).set_index('gene')\n",
    "    return result_df.sort_values('correlation', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83a31b9-6dee-4e76-bf7f-0bb11cd42b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_dfs = {}\n",
    "for ct in ['Arterial EC','Venous EC']:\n",
    "    ct_adata = endo_adata[endo_adata.obs['Cell Subtype']==ct]\n",
    "    df = correlate_genes_with_pseudotime(ct_adata,method='pearson',pseudotime='palantir_pseudotime')\n",
    "    corr_dfs[ct]=df.dropna(how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47d9a07-3669-4a2f-a92b-c480e832d6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_n_genes=50\n",
    "arterial_large_genes = corr_dfs['Arterial EC'].head(top_n_genes).index.tolist()\n",
    "venous_large_genes = corr_dfs['Venous EC'].head(top_n_genes).index.tolist()\n",
    "arterial_small_genes = corr_dfs['Arterial EC'].tail(top_n_genes).index.tolist()[::-1]\n",
    "venous_small_genes = corr_dfs['Venous EC'].tail(top_n_genes).index.tolist()[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d093c1f3-3e97-43e5-8413-93db1c27da6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2,venn3\n",
    "\n",
    "# Define your lists\n",
    "\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn = venn2([set(arterial_large_genes), set(venous_large_genes)], \n",
    "             set_labels=('Arterial', 'Venous'), \n",
    "             set_colors=('#4A90E2', '#E35D6A'), \n",
    "             alpha=0.7)\n",
    "\n",
    "# Optional: Customize font size\n",
    "for text in venn.set_labels:\n",
    "    text.set_fontsize(12)\n",
    "for text in venn.subset_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(12)\n",
    "\n",
    "# Show the plot\n",
    "plt.title(\"Top 50 genes positively correlated with pseudotime\")\n",
    "plt.savefig(f'{figures}/hurskainenvenn_diagram_large.png',dpi=300,bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b9f29fd-f468-4f62-8595-ab6bb5430268",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2\n",
    "\n",
    "# Define your lists\n",
    "\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn = venn2([set(arterial_small_genes), set(venous_small_genes)], \n",
    "             set_labels=('Arterial', 'Venous'), \n",
    "             set_colors=('#4A90E2', '#E35D6A'), \n",
    "             alpha=0.7)\n",
    "\n",
    "# Optional: Customize font size\n",
    "for text in venn.set_labels:\n",
    "    text.set_fontsize(12)\n",
    "for text in venn.subset_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(12)\n",
    "\n",
    "# Show the plot\n",
    "plt.title(\"Top 50 genes negatively correlated with pseudotime\")\n",
    "plt.savefig(f'{figures}/hurskainenvenn_diagram_small.png',dpi=300,bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3478a58-c6ec-48f7-b126-052174fa4f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_genes = [x for x in arterial_large_genes if x in venous_large_genes]\n",
    "small_genes = [x for x in arterial_small_genes if x in venous_small_genes]\n",
    "sc.tl.score_genes(endo_adata,large_genes,score_name='large_score')\n",
    "sc.tl.score_genes(endo_adata,small_genes,score_name='small_score')\n",
    "endo_adata.obs['Vessel size score'] = endo_adata.obs['large_score'] - endo_adata.obs['small_score']\n",
    "\n",
    "\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler()\n",
    "def normalize_dataframe(df):\n",
    "    # Initialize the MinMaxScaler\n",
    "    scaler = MinMaxScaler(feature_range=(0, 1))\n",
    "    # Fit the scaler on the data and transform each column\n",
    "    df_normalized = pd.DataFrame(scaler.fit_transform(df), index=df.index, columns=df.columns)\n",
    "    return df_normalized\n",
    "endo_adata.obs['Vessel size score'] = scaler.fit_transform(endo_adata.obs[['Vessel size score']])\n",
    "endo_adata.obs['Vessel size category'] = pd.cut(endo_adata.obs['Vessel size score'], bins=4,labels=['capillary','small','medium','large'])\n",
    "sc.pl.umap(endo_adata,color=['Vessel size score'],cmap='Oranges',size=size,frameon=False,save='_hurskainenvessel_size_score.png')\n",
    "sc.pl.umap(endo_adata,color=['Vessel size category'],cmap='viridis',size=size,frameon=False,save='_hurskainenvessel_size_category.png')\n",
    "sc.pl.umap(endo_adata,color=['Cell Subtype'],cmap='viridis',size=size,legend_loc='on data',legend_fontsize=10, legend_fontoutline=1,frameon=False,save='hurskainencellsubtype.png')\n",
    "sc.pl.umap(endo_adata,color=['Mgp'],cmap='viridis',size=size,frameon=False,save='hurskainenmgp.png')\n",
    "sc.pl.umap(endo_adata,color=['Col4a1'],cmap='viridis',size=size,frameon=False,save='hurskainencol4a1.png')\n",
    "sc.pl.umap(endo_adata,color=['Col4a2'],cmap='viridis',size=size,frameon=False,save='hurskainencol4a2.png')\n",
    "sc.pl.umap(endo_adata,color=['Eln'],cmap='viridis',size=size,frameon=False,save='hurskaineneln.png')\n",
    "sc.pl.umap(endo_adata,color=['Hey1'],cmap='viridis',size=size,frameon=False,save='hurskainenhey1.png')\n",
    "sc.pl.umap(endo_adata,color=['Nr2f2'],cmap='viridis',size=size,frameon=False,save='hurskainennr2f2.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb3864e-37ed-4ecf-850f-faafd1c30cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(endo_adata,color = large_genes + small_genes,cmap='viridis',hspace=0.5,save='hurskainen_allsize.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f72771-68fe-4422-a94c-46301f4e33f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_adata.write(f'{figures}/vessel_size.gz.h5ad',compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa9b945a-a258-4ceb-9cb3-25f732b0a998",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "def custom_dotplot(\n",
    "    adata, genes, x_obs, y_obs,\n",
    "    x_order=None, y_order=None,\n",
    "    min_expr=0.1, cmap='RdBu_r',\n",
    "    dot_max_size=300, pad=0.5,\n",
    "    scale_by_gene=False,\n",
    "    figsize=None,\n",
    "    save=None, dpi=300,\n",
    "    show_gridlines=False\n",
    "):\n",
    "    \"\"\"\n",
    "    Custom dotplot that mimics scanpy's style but allows:\n",
    "    - Custom x/y groupings\n",
    "    - Split x-axis for condition + gene\n",
    "    - Color = average expression, Size = percent expressing\n",
    "    - Optional scaling and exporting\n",
    "    \"\"\"\n",
    "    adata.obs[x_obs] = adata.obs[x_obs].astype('category')\n",
    "    adata.obs[y_obs] = adata.obs[y_obs].astype('category')\n",
    "\n",
    "    if x_order is None:\n",
    "        x_order = adata.obs[x_obs].cat.categories.tolist()\n",
    "    if y_order is None:\n",
    "        y_order = adata.obs[y_obs].cat.categories.tolist()\n",
    "\n",
    "    df = sc.get.obs_df(adata, keys=[x_obs, y_obs] + genes, layer=None)\n",
    "\n",
    "    results = []\n",
    "    for gene in genes:\n",
    "        for x_val in x_order:\n",
    "            for y_val in y_order:\n",
    "                group = df[(df[x_obs] == x_val) & (df[y_obs] == y_val)]\n",
    "                if group.shape[0] == 0:\n",
    "                    continue\n",
    "                expr = group[gene]\n",
    "                avg_expr = expr.mean()\n",
    "                prop_expr = (expr > min_expr).mean()\n",
    "                results.append({\n",
    "                    \"gene\": gene,\n",
    "                    \"x_group\": x_val,\n",
    "                    \"y_group\": y_val,\n",
    "                    \"avg_expr\": avg_expr,\n",
    "                    \"prop_expr\": prop_expr\n",
    "                })\n",
    "\n",
    "    plot_df = pd.DataFrame(results)\n",
    "    plot_df[\"x_label\"] = plot_df[\"gene\"] + \"\\n\" + plot_df[\"x_group\"]\n",
    "\n",
    "    # Scale avg_expr within gene\n",
    "    if scale_by_gene:\n",
    "        plot_df[\"scaled_expr\"] = plot_df.groupby(\"gene\")[\"avg_expr\"].transform(\n",
    "            lambda x: (x - x.min()) / (x.max() - x.min() + 1e-8)\n",
    "        )\n",
    "        color_col = \"scaled_expr\"\n",
    "    else:\n",
    "        color_col = \"avg_expr\"\n",
    "\n",
    "    # X/Y axis label arrangement\n",
    "    x_labels = []\n",
    "    x_groups = []\n",
    "    genes_list = []\n",
    "    for gene in genes:\n",
    "        for x_val in x_order:\n",
    "            x_labels.append(f\"{x_val}\\n{gene}\")\n",
    "            x_groups.append(x_val)\n",
    "            genes_list.append(gene)\n",
    "\n",
    "    y_labels = y_order\n",
    "\n",
    "    if figsize is None:\n",
    "        figsize = (len(x_labels) * 0.6 + 2, len(y_labels) * 0.6 + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # Scatter plot\n",
    "    for _, row in plot_df.iterrows():\n",
    "        x = x_labels.index(f\"{row['x_group']}\\n{row['gene']}\")\n",
    "        y = y_labels.index(row[\"y_group\"])\n",
    "        ax.scatter(\n",
    "            x, y,\n",
    "            s=row[\"prop_expr\"] * dot_max_size,\n",
    "            c=[row[color_col]],\n",
    "            cmap=cmap,\n",
    "            vmin=plot_df[color_col].min(),\n",
    "            vmax=plot_df[color_col].max(),\n",
    "            edgecolor='black',\n",
    "            linewidth=0.5\n",
    "        )\n",
    "\n",
    "    # Optional gridlines\n",
    "    if show_gridlines:\n",
    "        for y in range(len(y_labels)):\n",
    "            ax.axhline(y, color='lightgray', linestyle=':', linewidth=0.5)\n",
    "\n",
    "    # Vertical dashed lines between gene groups\n",
    "    for i in range(1, len(genes)):\n",
    "        xpos = i * len(x_order) - 0.5\n",
    "        ax.axvline(x=xpos, color='gray', linestyle='--', linewidth=1)\n",
    "\n",
    "    ax.set_xticks(range(len(x_labels)))\n",
    "    ax.set_xticklabels(x_groups, rotation=0, ha='center')\n",
    "    ax.set_yticks(range(len(y_labels)))\n",
    "    ax.set_yticklabels(y_labels)\n",
    "    ax.set_xlim(-pad, len(x_labels) - 1 + pad)\n",
    "    ax.set_ylim(-pad, len(y_labels) - 1 + pad)\n",
    "    ax.invert_yaxis()\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel(y_obs)\n",
    "\n",
    "    # # Add gene names on second x-axis\n",
    "    ax_gene = ax.secondary_xaxis('top')\n",
    "    gene_locs = range(len(x_labels))\n",
    "    gene_locs = [(gene_locs[i] + gene_locs[i + 1]) / 2 for i in range(len(gene_locs) - 1)][::2]\n",
    "    ax_gene.set_xticks(gene_locs)\n",
    "    # if len(set(genes_list)) == 1:\n",
    "    #     ax_gene.set_xticklabels(genes_list[::2], rotation=0, ha='center')\n",
    "    # else:\n",
    "    ax_gene.set_xticklabels(genes_list[::2], rotation=0, ha='center')\n",
    "    # # ax_gene.set_xlabel(\"Gene\")\n",
    "\n",
    "    # Colorbar\n",
    "    norm = plt.Normalize(plot_df[color_col].min(), plot_df[color_col].max())\n",
    "    sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "    sm.set_array([])\n",
    "    cbar = fig.colorbar(sm, ax=ax, location='right', pad=0.02)\n",
    "    cbar.set_label('Mean expression\\nin group')\n",
    "\n",
    "    # Dot size legend\n",
    "    prop_vals = plot_df[\"prop_expr\"]\n",
    "    min_pct = int(np.floor(prop_vals.min() * 100 / 5) * 5)\n",
    "    max_pct = int(np.ceil(prop_vals.max() * 100 / 5) * 5)\n",
    "    possible_labels = [i for i in range(min_pct, max_pct + 1) if i % 5 == 0]\n",
    "    num_labels = min(4, len(possible_labels))\n",
    "    legend_labels = np.linspace(min_pct, max_pct, num_labels, dtype=int)\n",
    "\n",
    "    handles = [\n",
    "        plt.scatter([], [], s=(pct / 100) * dot_max_size, c='gray',\n",
    "                    edgecolors='black', linewidth=0.5, label=f\"{pct}%\")\n",
    "        for pct in legend_labels\n",
    "    ]\n",
    "\n",
    "    # Move axis to left to make room for both legends\n",
    "    box = ax.get_position()\n",
    "    ax.set_position([box.x0, box.y0, box.width * 0.72, box.height])\n",
    "\n",
    "        # --- Dot size legend with clean rounding ---\n",
    "    min_pct_raw = plot_df[\"prop_expr\"].min() * 100\n",
    "    max_pct_raw = plot_df[\"prop_expr\"].max() * 100\n",
    "    \n",
    "    min_pct = int(np.floor(min_pct_raw / 5) * 5)\n",
    "    max_pct = int(np.ceil(max_pct_raw / 5) * 5)\n",
    "    \n",
    "    # Generate 4 nicely rounded ticks between min and max\n",
    "    if max_pct - min_pct < 15:\n",
    "        ticks = np.linspace(min_pct, max_pct, 4)\n",
    "    else:\n",
    "        ticks = np.round(np.linspace(min_pct, max_pct, 4) / 5) * 5\n",
    "    \n",
    "    ticks = np.clip(ticks, 0, 100).astype(int)\n",
    "    \n",
    "    handles = [\n",
    "        plt.scatter([], [], s=(p / 100) * dot_max_size, c='gray', edgecolors='black')\n",
    "        for p in ticks\n",
    "    ]\n",
    "    labels = [f\"{p}%\" for p in ticks]\n",
    "    \n",
    "    fig.legend(\n",
    "        handles,\n",
    "        labels,\n",
    "        title=\"Pct. Expressing\",\n",
    "        loc='center right',\n",
    "        bbox_to_anchor=(1.40, 0.6),\n",
    "        frameon=True\n",
    "    )\n",
    "\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if save:\n",
    "        plt.savefig(save, dpi=dpi, bbox_inches='tight')\n",
    "        print(f\"Saved to {save}\")\n",
    "    else:\n",
    "        plt.show()\n",
    "adata_ven = endo_adata[endo_adata.obs['Cell Subtype']=='Venous EC']\n",
    "adata_ven.obs['proliferation_score'] = normalize_dataframe(adata.obs[['proliferation_score']])\n",
    "df = sc.get.obs_df(adata_ven,['proliferation_score','Vessel size category', 'Treatment','Timepoint'])\n",
    "df.rename(columns={'proliferation_score':'Proliferation score'},inplace=True)\n",
    "df = df.loc[df['Vessel size category']!= 'capillary']\n",
    "for tp in endo_adata.obs['Timepoint'].cat.categories:\n",
    "    fig, axs = plt.subplots(1, 2,figsize=(3,2),sharey=True)\n",
    "    axs = axs.ravel()\n",
    "    hue_order = ['small', 'medium', 'large']\n",
    "    palette = endo_adata.uns['Vessel size category_colors'][1:]\n",
    "    \n",
    "    for i, treat in enumerate(['Normoxia', 'Hyperoxia']):\n",
    "        ax = sns.kdeplot(\n",
    "            data=df[(df['Treatment'] == treat)&(df['Timepoint'] == tp)],\n",
    "            x=\"Proliferation score\",\n",
    "            hue='Vessel size category',\n",
    "            hue_order=hue_order,\n",
    "            palette=palette,\n",
    "            common_norm=False,\n",
    "            ax=axs[i]\n",
    "        )\n",
    "    \n",
    "        ax.set_title(treat)\n",
    "        ax.set_ylabel('Proportion')\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_xticks([0,1])\n",
    "        ax.set_xticklabels(['0', '1'])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "    \n",
    "        if i == 0:\n",
    "            # ✅ Line-style legend handles in reverse order\n",
    "            handles = [\n",
    "                mlines.Line2D([], [], color=palette[j], label=label, linewidth=2)\n",
    "                for j, label in reversed(list(enumerate(hue_order)))\n",
    "            ]\n",
    "            ax.legend(handles=handles[::-1], frameon=False, fontsize=\"8\", title='', loc=\"upper right\")\n",
    "        else:\n",
    "            ax.get_legend().remove()\n",
    "    \n",
    "    fig.supxlabel('Proliferation score', y=0.15, x=0.52)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(f'{figures}/histplot_venous_ec_treat_proliferation_score_{tp}.png', dpi=300, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    adata_ven_tp= adata_ven[adata_ven.obs['Timepoint']==tp]\n",
    "    adata_ven_tp.obs['Treatment'] = adata_ven_tp.obs['Treatment'].str[0]\n",
    "    custom_dotplot(adata_ven_tp[adata_ven_tp.obs['Vessel size category']!='capillary'],\n",
    "                   ['Mki67', 'Top2a', 'Birc5', 'Hmgb2', 'Cenpf'],\n",
    "                   scale_by_gene=True, \n",
    "                   x_obs='Treatment',\n",
    "                   y_obs='Vessel size category',\n",
    "                   x_order=['N','H'],\n",
    "                   cmap='Reds',\n",
    "                   dot_max_size=100,\n",
    "                   figsize=(4,2),\n",
    "                   save=f'{figures}/dotplot_proliferation_vec_{tp}.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "859a130e-6ce0-4e53-a056-2a27c30cf965",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, axs = plt.subplots(1, 2,figsize=(3,2),sharey=True)\n",
    "axs = axs.ravel()\n",
    "hue_order = ['P3', 'P7', 'P14']\n",
    "palette = endo_adata.uns['Timepoint_colors']\n",
    "df = adata_ven.obs[['proliferation_score','Timepoint','Treatment']]\n",
    "for i, treat in enumerate(['Normoxia', 'Hyperoxia']):\n",
    "    ax = sns.kdeplot(\n",
    "        data=df.loc[(df['Treatment'] == treat)],\n",
    "        x=\"proliferation_score\",\n",
    "        hue='Timepoint',\n",
    "        hue_order=hue_order,\n",
    "        palette=palette,\n",
    "        common_norm=False,\n",
    "        ax=axs[i]\n",
    "    )\n",
    "\n",
    "    ax.set_title(treat)\n",
    "    ax.set_ylabel('Proportion')\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_xticks([0,1])\n",
    "    ax.set_xticklabels(['0', '1'])\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "    if i == 0:\n",
    "        # ✅ Line-style legend handles in reverse order\n",
    "        handles = [\n",
    "            mlines.Line2D([], [], color=palette[j], label=label, linewidth=2)\n",
    "            for j, label in reversed(list(enumerate(hue_order)))\n",
    "        ]\n",
    "        ax.legend(handles=handles[::-1], frameon=False, fontsize=\"8\", title='', loc=\"upper right\")\n",
    "    else:\n",
    "        ax.get_legend().remove()\n",
    "\n",
    "fig.supxlabel('Proliferation score', y=0.15, x=0.52)\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/histplot_venous_ec_timepoint_proliferation.png', dpi=300, bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d4cffc-b365-4579-9d64-d5a87013b3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1,1,figsize=(2,2))\n",
    "palantir.plot.plot_trajectory(\n",
    "    endo_adata, # your anndata\n",
    "    \"Arterial EC\", # the branch to plot\n",
    "    cell_color=\"dpt_pseudotime\", # the ad.obs colum to color the cells by\n",
    "    n_arrows=5, # the number of arrow heads along the path\n",
    "    color='#4A90E2', # the color of the path and arrow heads\n",
    "    scanpy_kwargs=dict(cmap=\"viridis\",size=size), # arguments passed to scanpy.pl.embedding\n",
    "    arrowprops=dict(arrowstyle=\"->,head_length=.25,head_width=.25\", lw=2), # appearance of the arrow heads\n",
    "    lw=2, # thickness of the path\n",
    "ax=ax\n",
    "    # pseudotime_interval=(0, .9), # interval of the pseudotime to cover with the path\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(f'{figures}/hurskainenpalantir_art_trajectory.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a8b44ed-91b8-4545-954d-daa9e41548d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1,1,figsize=(2,2))\n",
    "palantir.plot.plot_trajectory(\n",
    "    endo_adata, # your anndata\n",
    "    \"Venous EC\", # the branch to plot\n",
    "    cell_color=\"dpt_pseudotime\", # the ad.obs colum to color the cells by\n",
    "    n_arrows=5, # the number of arrow heads along the path\n",
    "    color='#E35D6A', # the color of the path and arrow heads\n",
    "    scanpy_kwargs=dict(cmap=\"viridis\",size=size), # arguments passed to scanpy.pl.embedding\n",
    "    arrowprops=dict(arrowstyle=\"->,head_length=.25,head_width=.25\", lw=2), # appearance of the arrow heads\n",
    "    lw=2, # thickness of the path\n",
    "ax=ax\n",
    "    # pseudotime_interval=(0, .9), # interval of the pseudotime to cover with the path\n",
    ")\n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{figures}/hurskainenpalantir_ven_trajectory.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c30fc1c-2640-48ca-a109-6ba2d473d383",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for ct in ['Arterial EC','Venous EC']:\n",
    "    df = sc.get.obs_df(endo_adata,['Treatment','Timepoint','Cell Subtype','Vessel size score'])\n",
    "    df_ct = df.loc[df['Cell Subtype'] == ct]\n",
    "    for tp in ['P3','P7','P14']:\n",
    "        df = df_ct.loc[df_ct['Timepoint']==tp]\n",
    "        fig, ax = plt.subplots(1, figsize=(3,2),sharey=True)\n",
    "        hue_order = ['Normoxia','Hyperoxia']\n",
    "        # palette =['blue','red'],\n",
    "    \n",
    "        ax = sns.kdeplot(\n",
    "            data=df,\n",
    "            x=\"Vessel size score\",\n",
    "            hue='Treatment',\n",
    "            hue_order=hue_order,\n",
    "            cut=0,\n",
    "            # palette=palette,\n",
    "            common_norm=False,\n",
    "            # stat='probability',\n",
    "            # element='poly',\n",
    "            # fill=False,\n",
    "            # common_norm=False,\n",
    "            # bins=20,\n",
    "        )\n",
    "    \n",
    "        ax.set_title(f'{ct}\\nVessel size distribution')\n",
    "        ax.set_ylabel('Proportion')\n",
    "        # ax.set_ylim([0,0.1])\n",
    "        # ax.set_xlim([0.2,1])\n",
    "        # ax.set_xticklabels([])\n",
    "        # ax.set_xlabel('')\n",
    "        # ax.get_legend().remove()\n",
    "        \n",
    "        # fig.supxlabel('Vessel size score\\n0-1', y=0.15, x=0.52)\n",
    "        sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(f'{figures}/histplot_vessel_size_disease_{ct}_{tp}.png', dpi=300, bbox_inches='tight')\n",
    "        plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb79c80a-9722-4d1f-8e6c-ba93fb88759e",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13f75cf7-7513-4382-9687-b6549d3bfc3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(endo_adata,color=['Vessel size score'],cmap='viridis',size=size,frameon=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b9d2b80-f152-48c7-bfac-f99f23b9c958",
   "metadata": {},
   "outputs": [],
   "source": [
    "for tp in ['P3','P7','P14']:\n",
    "    tp_adata = endo_adata[endo_adata.obs['Timepoint']==tp]\n",
    "    tp_adata.X = tp_adata.layers['cc_regress']\n",
    "    figures_tp = f\"data/figures/figures/hurskainen/{tp}\"\n",
    "    os.makedirs(figures_tp, exist_ok=True)\n",
    "    sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "    sc.settings.figdir = figures_tp\n",
    "    sc.pp.highly_variable_genes(tp_adata,batch_key='orig.ident')\n",
    "    sc.pp.pca(tp_adata, mask_var=\"highly_variable\")\n",
    "    sce.pp.harmony_integrate(tp_adata,'orig.ident',max_iter_harmony = 20)\n",
    "    sc.pp.neighbors(tp_adata, use_rep=\"X_pca\")\n",
    "    sc.tl.leiden(tp_adata, key_added=\"leiden\")\n",
    "    sc.tl.umap(tp_adata,min_dist=1)\n",
    "    tp_adata.X = tp_adata.layers['log1p']\n",
    "    sc.tl.rank_genes_groups(tp_adata,'leiden',method='wilcoxon')\n",
    "    sc.pl.rank_genes_groups_dotplot(tp_adata,dendrogram=False)\n",
    "    for color in ['celltype','Cell Subtype','Timepoint','Treatment','leiden','Gja5','Slc6a2','Kit','Hey1','Nr2f2','Mecom','Eln','Mgp','Col4a1','Col4a2']:\n",
    "        sc.pl.umap(tp_adata,color=color,use_raw=False,cmap='viridis',size=size,frameon=False,save=f'{tp}_{color}.png')\n",
    "    for treat in ['Normoxia','Hyperoxia']:\n",
    "        treat_adata = tp_adata[tp_adata.obs['Treatment']==treat]\n",
    "        treat_adata.X = treat_adata.layers['cc_regress']\n",
    "        figures_tp = f\"data/figures/figures/hurskainen/{tp}/{treat}\"\n",
    "        os.makedirs(figures_tp, exist_ok=True)\n",
    "        sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "        sc.settings.figdir = figures_tp\n",
    "        sc.pp.highly_variable_genes(treat_adata,batch_key='orig.ident')\n",
    "        sc.pp.pca(treat_adata, mask_var=\"highly_variable\")\n",
    "        sce.pp.harmony_integrate(treat_adata,'orig.ident',max_iter_harmony = 20)\n",
    "        sc.pp.neighbors(treat_adata, use_rep=\"X_pca\")\n",
    "        sc.tl.leiden(treat_adata, key_added=\"leiden\")\n",
    "    \n",
    "        sc.tl.umap(treat_adata,min_dist=1)\n",
    "        treat_adata.X = treat_adata.layers['log1p']\n",
    "        sc.tl.rank_genes_groups(treat_adata,'leiden',method='wilcoxon')\n",
    "        sc.pl.rank_genes_groups_dotplot(treat_adata,dendrogram=False)\n",
    "        for color in ['celltype','Cell Subtype','Timepoint','Treatment','leiden','Gja5','Slc6a2','Kit','Hey1','Nr2f2','Mecom','Eln','Mgp','Col4a1','Col4a2']:\n",
    "            sc.pl.umap(treat_adata,color=color,use_raw=False,cmap='viridis',size=size,frameon=False,save=f'{tp}_{color}.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c391d557-804a-4d58-a1c4-85da4a2d6986",
   "metadata": {},
   "outputs": [],
   "source": [
    "for treat in ['Normoxia','Hyperoxia']:\n",
    "    treat_adata = endo_adata[endo_adata.obs['Treatment']==treat]\n",
    "    treat_adata.X = treat_adata.layers['cc_regress']\n",
    "    figures_tp = f\"data/figures/figures/hurskainen/{treat}\"\n",
    "    os.makedirs(figures_tp, exist_ok=True)\n",
    "    sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "    sc.settings.figdir = figures_tp\n",
    "    sc.pp.highly_variable_genes(treat_adata,batch_key='orig.ident')\n",
    "    sc.pp.pca(treat_adata, mask_var=\"highly_variable\")\n",
    "    sce.pp.harmony_integrate(treat_adata,'orig.ident',max_iter_harmony = 20)\n",
    "    sc.pp.neighbors(treat_adata, use_rep=\"X_pca\")\n",
    "    sc.tl.leiden(treat_adata, key_added=\"leiden\")\n",
    "\n",
    "    sc.tl.umap(treat_adata,min_dist=1)\n",
    "    tp_adata.X = tp_adata.layers['log1p']\n",
    "    sc.tl.rank_genes_groups(treat_adata,'leiden',method='wilcoxon')\n",
    "    sc.pl.rank_genes_groups_dotplot(treat_adata,dendrogram=False)\n",
    "    for color in ['celltype','Cell Subtype','Timepoint','Treatment','leiden','Gja5','Slc6a2','Kit','Hey1','Nr2f2','Mecom','Eln','Mgp','Col4a1','Col4a2']:\n",
    "        sc.pl.umap(treat_adata,color=color,use_raw=False,cmap='viridis',size=size,frameon=False,save=f'{tp}_{color}.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ef2f600-8f6c-44ee-a5b9-ec3bf6f5aa97",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa66419-79d4-484f-8f29-2b186a145e11",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = endo_adata.obs['Cell Subtype'] =='Venous EC'\n",
    "sc.pl.umap(endo_adata.copy(),color=['Timepoint'],mask_obs=mask,size=size,na_in_legend=False,frameon=False,save='hurskainen_ven_timepoint.png')\n",
    "sc.pl.umap(endo_adata.copy(),color=['Treatment'],palette=endo_adata.uns['Treatment_colors'][::-1],mask_obs=mask,size=size,na_in_legend=False,frameon=False,save='hurskainen_ven_treatment.png')\n",
    "sc.pl.umap(endo_adata.copy(),color=['Vessel size category'],mask_obs=mask,size=size,na_in_legend=False,frameon=False,save='hurskainen_ven_size.png')\n",
    "sc.pl.umap(endo_adata.copy(),color=['Mki67'],cmap='viridis',mask_obs=mask,size=size,na_in_legend=False,frameon=False,save='hurskainen_ven_mki67.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b52cdc-0f18-449d-9249-d04bf30c5cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1,1))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15\n",
    "for gene in ['Cxcl12','Cxcr4','Ackr3','Esr2','Dkk2','Moxd1']:\n",
    "    sc.pl.umap(endo_adata,color=[gene],cmap='viridis',size=size,frameon=False,save=f'hurskainen_{gene}.png')\n",
    "os.makedirs(figures, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "sc.settings.figdir = figures\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "size=15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3c93955-724e-4bc9-81ad-b9fc4bd4a224",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls = []\n",
    "for x,y in zip(endo_adata.obs['Cell Subtype'],endo_adata.obs['Vessel size category']):\n",
    "    if x =='Cap1':\n",
    "        ls.append('Cap1')\n",
    "    else:\n",
    "        if y =='capillary':\n",
    "            ls.append(f'Cap1')\n",
    "            continue\n",
    "        if x =='Arterial EC':\n",
    "            x = 'PAEC'\n",
    "        else:\n",
    "            x ='PVEC'\n",
    "        ls.append(f'{x} {y[0].upper()}')\n",
    "endo_adata.obs['ct_s'] = ls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48e54e20-b630-488c-9857-3e4aa6b9c4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi_save=300, fontsize=10, figsize=(1.5,1.5))\n",
    "\n",
    "gene_ls = [\n",
    "    'Ccdc85a',\n",
    "    'Glp1r',\n",
    "    'Kit',\n",
    "         'Sox4','Nrp1','Ifitm3',\n",
    "         'Ptprr','Adgrg6','Foxo1',\n",
    "         'Mgp','Eln','Nr4a2',\n",
    "][::-1]\n",
    "sc.pl.dotplot(endo_adata,gene_ls,groupby='ct_s',cmap='viridis',categories_order=['PAEC L','PAEC M','PAEC S','Cap1','PVEC S','PVEC M', 'PVEC L'],\n",
    "              standard_scale='var',save='size_axis.png')\n",
    "sc.pl.umap(endo_adata,color=gene_ls,hspace=0.5,wspace=0.3,cmap='viridis',save='size_axis.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fabd8a4-77eb-47c9-b8ad-fceee5d737cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ls = ['Dkk2','Sox6',\n",
    "           'Lama3','Bdkrb2',\n",
    "           # 'Depp1', not present\n",
    "           'Stc1',\n",
    "\n",
    "           'Adam23','Ntrk2',\n",
    "           'Emid1','Chrm2',\n",
    "           'Chrm3','Rarb',\n",
    "           # 'Fads2b', not present\n",
    "           'Gria3',\n",
    "           'Ptger3','Moxd1',\n",
    "             ]\n",
    "sc.pl.MatrixPlot(endo_adata,gene_ls,groupby='ct_s',categories_order=['PAEC L','PAEC M','PAEC S','Cap1','PVEC S','PVEC M', 'PVEC L'],\n",
    "              standard_scale='var').style(cmap='viridis').swap_axes().savefig(f'{figures}/matrixplot_size_markers.png',dpi=300,bbox_inches='tight')\n",
    "sc.pl.DotPlot(endo_adata,gene_ls,groupby='ct_s',categories_order=['PAEC L','PAEC M','PAEC S','Cap1','PVEC S','PVEC M', 'PVEC L'],\n",
    "              standard_scale='var').style(cmap='viridis').savefig(f'{figures}/matrixplot_size_markers.png',dpi=300,bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
